### PR TITLE
[Feature] Support schema changes to the lake table of primary key

### PR DESCRIPTION
**Pull request by sduzh**
## What type of PR is this：
- [ ] BugFix
- [x] Feature
- [ ] Enhancement
- [ ] Refactor
- [ ] UT
- [ ] Doc
- [ ] Tool

## Which issues of this PR fixes ：

Fixes #

## Checklist:

- [x] I have added test cases for my bug fix or my new feature
- [ ] This pr will affect users' behaviors
- [ ] This pr needs user documentation (for new or modified features or behaviors)
  - [ ] I have added documentation for my new feature or new function

## Bugfix cherry-pick branch check:
- [ ] I have checked the version labels which the pr will be auto backported to target branch
  - [ ] 2.5
  - [ ] 2.4
  - [ ] 2.3
  - [ ] 2.2

### CODE REVIEW

**Comment by wanpengfei-git** [comment]
#### [FE PR Coverage Check]


:heart_eyes: **pass** : 0 / 0 (0%)

**Comment by wanpengfei-git** [comment]
#### [FE PR Coverage Check]


:heart_eyes: **pass** : 0 / 0 (0%)

**Comment by wanpengfei-git** [comment]
#### [FE PR Coverage Check]


:heart_eyes: **pass** : 0 / 0 (0%)

**Comment by wanpengfei-git** [comment]
#### [FE PR Coverage Check]


:heart_eyes: **pass** : 0 / 0 (0%)

**Comment by wanpengfei-git** [comment]
#### [FE PR Coverage Check]


:heart_eyes: **pass** : 0 / 0 (0%)

**Comment by wanpengfei-git** [comment]
#### [FE PR Coverage Check]


:heart_eyes: **pass** : 0 / 0 (0%)

File: be/src/storage/lake/tablet_manager.cpp [hunk 1]
```diff
@@ -460,181 +462,11 @@ StatusOr<TabletSchemaPtr> TabletManager::get_tablet_schema(int64_t tablet_id) {
 StatusOr<double> TabletManager::publish_version(int64_t tablet_id, int64_t base_version, int64_t new_version,
                                                 const int64_t* txns, int txns_size) {
     ASSIGN_OR_RETURN(auto tablet, get_tablet(tablet_id));
-    return publish(_location_provider, &tablet, base_version, new_version, txns, txns_size);
-}
-
-static Status apply_write_log(const TxnLogPB_OpWrite& op_write, TabletMetadata* metadata) {
```
Comment (sduzh) [inline be/src/storage/lake/tablet_manager.cpp:466]: Moved to `NonPrimaryKeyTxnLogApplier::apply_write_log()`

File: be/src/storage/lake/tablet_manager.cpp [hunk 2]
```diff
@@ -460,181 +462,11 @@ StatusOr<TabletSchemaPtr> TabletManager::get_tablet_schema(int64_t tablet_id) {
...
-        } else {
-            // delete
-            metadata->set_next_rowset_id(metadata->next_rowset_id() + 1);
-        }
-    }
-    return Status::OK();
+    return publish(&tablet, base_version, new_version, txns, txns_size);
 }
 
-static Status apply_pk_write_log(const TxnLogPB_OpWrite& op_write, const TabletMetadata& metadata, Tablet* tablet,
```
Comment (sduzh) [inline be/src/storage/lake/tablet_manager.cpp:481]: Moved to `PrimaryKeyTxnLogApplier::apply_write_log()`

File: be/src/storage/lake/tablet_manager.cpp [hunk 3]
```diff
@@ -460,181 +462,11 @@ StatusOr<TabletSchemaPtr> TabletManager::get_tablet_schema(int64_t tablet_id) {
...
-
-    auto st = tablet->update_mgr()->publish_primary_compaction(op_compaction, metadata, tablet, builder, base_version);
-    if (!st.ok()) {
-        LOG(WARNING) << "Fail to apply_pk_compaction_log: " << tablet->id() << " st: " << st;
-        builder->handle_failure();
-    }
-    return st;
-}
-
-static Status apply_compaction_log(const TxnLogPB_OpCompaction& op_compaction, TabletMetadata* metadata) {
```
Comment (sduzh) [inline be/src/storage/lake/tablet_manager.cpp:511]: Moved to `NonPrimaryKeyTxnLogApplier::apply_compaction_log()`

File: be/src/storage/lake/meta_file.h [hunk 4]
```diff
@@ -33,26 +34,26 @@ class UpdateManager;
...
-    Status _finalize_delvec(LocationProvider* location_provider, int64_t version);
+    Status _finalize_delvec(int64_t version);
 
 private:
+    Tablet _tablet;
     std::shared_ptr<TabletMetadata> _tablet_meta;
     UpdateManager* _update_mgr;
     Buffer<uint8_t> _buf;
     std::unordered_map<uint32_t, DelvecPairPB> _delvecs;
-    std::vector<std::pair<TabletSegmentId, DelVectorPtr>> _cache_delvec_updates;
```
Comment (luohaha) [inline be/src/storage/lake/meta_file.h:55]: Why remove this, delvec is small and it is essential for reading latency.

File: be/test/CMakeLists.txt [hunk 5]
```diff
@@ -186,7 +186,6 @@ set(EXEC_FILES
         ./storage/lake/tablet_reader_test.cpp
         ./storage/lake/tablet_writer_test.cpp
         ./storage/lake/primary_key_compaction_task_test.cpp
-        ./storage/lake/primary_key_test.cpp
```
Comment (luohaha) [inline be/test/CMakeLists.txt:189]: Why delete this?

File: be/src/storage/lake/update_manager.h [hunk 6]
```diff
@@ -114,15 +111,7 @@ class UpdateManager {
 
     // rowset cache
     DynamicCache<string, RowsetUpdateState> _update_state_cache;
-
-    // DelVector related states
-    std::mutex _del_vec_cache_lock;
-    std::map<TabletSegmentId, DelVectorPtr> _del_vec_cache;
```
Comment (luohaha) [inline be/src/storage/lake/update_manager.h:120]: And this?

File: be/src/storage/lake/txn_log_applier.cpp [hunk 7]
```diff
@@ -0,0 +1,270 @@
...
+    Status apply_schema_change_log(const TxnLogPB_OpSchemaChange& op_schema_change) {
+        for (const auto& rowset : op_schema_change.rowsets()) {
+            auto new_rowset = _metadata->add_rowsets();
+            new_rowset->CopyFrom(rowset);
+            new_rowset->set_id(_metadata->next_rowset_id());
+            _metadata->set_next_rowset_id(_metadata->next_rowset_id() + std::max(1, new_rowset->segments_size()));
+        }
+        if (op_schema_change.has_delvec_meta()) {
+            _metadata->mutable_delvec_meta()->CopyFrom(op_schema_change.delvec_meta());
+        }
```
Comment (wyb) [inline be/src/storage/lake/txn_log_applier.cpp:254]: non primary key table does not have delvec meta

File: be/src/storage/lake/schema_change.cpp [hunk 8]
```diff
@@ -364,6 +374,16 @@ Status SchemaChangeHandler::convert_historical_rowsets(const SchemaChangeParams&
         }
     }
 
+    // copy delete vector files if necessary
+    if (op_schema_change->linked_segment() && base_metadata->has_delvec_meta()) {
+        for (const auto& delvec : base_metadata->delvec_meta().delvecs()) {
+            auto src = base_tablet->del_location(tablet_delvec_filename(base_tablet->id(), delvec.page().version()));
+            auto dst = new_tablet->del_location(tablet_delvec_filename(new_tablet->id(), delvec.page().version()));
```
[the remark this answers is not in the document]
Comment (sduzh) [reply be/src/storage/lake/schema_change.cpp]: Updated, PTAL

File: be/src/storage/lake/txn_log_applier.cpp (hunk 7 above)
Comment (sduzh) [reply be/src/storage/lake/txn_log_applier.cpp]: Updated, PTAL

File: be/test/CMakeLists.txt (hunk 5 above)
Comment (sduzh) [reply be/test/CMakeLists.txt:189]: Merged into `tablet_writer_test.cpp`

**Comment by wanpengfei-git** [comment]
#### [FE PR Coverage Check]


:heart_eyes: **pass** : 0 / 0 (0%)

**Comment by wanpengfei-git** [comment]
#### [FE PR Coverage Check]


:heart_eyes: **pass** : 0 / 0 (0%)

**Review by luohaha — APPROVED** [review]

**Review by wyb — APPROVED** [review]

**Comment by wanpengfei-git** [comment]
run starrocks_admit_test

**Review by Seaven — APPROVED** [review]

**Comment by sduzh** [comment]
run starrocks_admit_test

**Pull request merged**